cumulusci/tasks/apex/testrunner.py

```diff
@@ -0,0 +1,339 @@
+
+        self.counts = {}
+
+    # pylint: disable=W0201
+    def _init_class(self):
+        self.classes_by_id = {}
+        self.classes_by_name = {}
+        self.job_id = None
+        self.results_by_class_name = {}
+        self.result = None
+
+    def _decode_to_unicode(self, content):
+        if content:
+            try:
+                # Try to decode ISO-8859-1 to unicode
+                return content.decode('ISO-8859-1')
+            except UnicodeEncodeError:
+                # Assume content is unicode already
+                return content
+
+    def _get_test_classes(self):
+        if self.options['managed']:
+            namespace = self.options.get('namespace')
+            if not namespace:
+                raise TaskOptionsError(
+                    'Running tests in managed mode but no namespace available.'
+                )
+            namespace = "'{}'".format(namespace)
+        else:
+            namespace = 'null'
+        # Split by commas to allow multiple class name matching options
+        test_name_match = self.options['test_name_match']
+        included_tests = []
+        for pattern in test_name_match.split(','):
+            if pattern:
+                included_tests.append("Name LIKE '{}'".format(pattern))
+        # Add any excludes to the where clause
+        test_name_exclude = self.options.get('test_name_exclude', '')
+        excluded_tests = []
+        for pattern in test_name_exclude.split(','):
+            if pattern:
+                excluded_tests.append("(NOT Name LIKE '{}')".format(pattern))
+        # Get all test classes for namespace
+        query = ('SELECT Id, Name FROM ApexClass ' +
+                 'WHERE NamespacePrefix = {}'.format(namespace))
+        if included_tests:
+            query += ' AND ({})'.format(' OR '.join(included_tests))
+        if excluded_tests:
+            query += ' AND {}'.format(' AND '.join(excluded_tests))
+        # Run the query
+        self.logger.info('Running query: {}'.format(query))
+        result = self.tooling.query_all(query)
+        self.logger.info('Found {} test classes'.format(result['totalSize']))
+        return result
+
+    def _get_test_results(self):
+        result = self.tooling.query_all(TEST_RESULT_QUERY.format(self.job_id))
+        self.counts = {
+            'Pass': 0,
+            'Fail': 0,
+            'CompileFail': 0,
+            'Skip': 0,
+        }
+        for test_result in result['records']:
+            class_name = self.classes_by_id[test_result['ApexClassId']]
+            self.results_by_class_name[class_name][test_result[
+                'MethodName']] = test_result
+            self.counts[test_result['Outcome']] += 1
+        test_results = []
+        class_names = self.results_by_class_name.keys()
+        class_names.sort()
+        for class_name in class_names:
+            message = 'Class: {}'.format(class_name)
+            duration = None  # TODO: Get class total runtime
```

Comment:
that's what i get for checking in my TODO. it was only logged to the console, so its safe to lose.